mypy/checkexpr.py

```diff
@@ -2650,9 +2650,10 @@ def overload_arg_similarity(actual: Type, formal: Type) -> int:
             # NoneTyp matches anything if we're not doing strict Optional checking
             return 2
         else:
-            # NoneType is a subtype of object
+            # NoneType can be promoted to object, but isn't actually a subtype (otherwise you
+            # wouldn't be able to define overloads between object and None).
             if isinstance(formal, Instance) and formal.type.fullname() == "builtins.object":
-                return 2
+                return 1
```

Comment:
Update the docstring above to mention that we treat `None` -> `object` as a promotion though it's actually treated as a real subtype elsewhere.

Reply:
Oh, I forgot that this was true. That makes this seem more hacky.

A general solution (not specific to None or strict-optional) would be to add a fourth level of overload arg similarity for subclasses, e.g.

    0: No match
    1: Actual can be promoted to formal
    2: Actual is a subclass of formal
    3: Actual is the same as formal

Would that be preferable? I'd be happy to implement (optionally including refactoring this function to return an `Enum`) if so. It seems less hacky but much more likely to have unforseen consequences/introduce new errors.

Reply:
TBH, I was also thinking that three levels of similarity might be not enough, see e.g. https://github.com/python/mypy/issues/3656. But introducing a new level of similarity is a big change, so that I think we should wait until @JukkaL is back.

Reply:
Treating promotions as different from other forms of compatibility is already unsafe, and adding more levels of similarity would add more room for unsafety. I'd prefer to use the hack since its effect is limited. Basically we'd only surgically enable a specific use case that we know to be important, but otherwise overload resolution would remain the same.